File: debot/src/main/java/com/tomoima/debot/strategy/DebotCallActivityMethodStrategy.java
package com.tomoima.debot.strategy;


import android.app.Activity;
import android.support.annotation.NonNull;

import com.tomoima.debot.annotation.DebotAnnotation;

import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;


public class DebotCallActivityMethodStrategy extends DebotStrategy{

    final String methodName;

    /**
     * Constructor. A parameter should be same as a method name you want to call from Activity
     * @param methodName method name to call
     */
    public DebotCallActivityMethodStrategy(@NonNull String methodName){
        this.methodName = methodName;
    }

    @Override
    public void startAction(@NonNull Activity activity) {
        Class clazz = activity.getClass();
        Method[] methods = clazz.getDeclaredMethods();
        for (Method method : methods) {
            if (!method.isAnnotationPresent(DebotAnnotation.class)) continue;
            Annotation annotation = method.getAnnotation(DebotAnnotation.class);
            if (methodName.equals(((DebotAnnotation) annotation).value())){
                try {
                    method.invoke(activity);
                } catch (InvocationTargetException e) {
                    e.printStackTrace();
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                }
            }

        }
    }
}

File: debot/src/main/java/com/tomoima/debot/strategy/DebotCallActivityMethodStrategy.kt
package com.tomoima.debot.strategy


import android.app.Activity
import com.tomoima.debot.annotation.DebotAnnotation
import java.lang.reflect.InvocationTargetException


class DebotCallActivityMethodStrategy
/**
 * Constructor. A parameter should be same as a method name you want to call from Activity
 * @param methodName method name to call
 */
(internal val methodName: String) : DebotStrategy() {

    override fun startAction(activity: Activity) {
        val clazz = activity.javaClass
        val methods = clazz.declaredMethods
        for (method in methods) {
            if (!method.isAnnotationPresent(DebotAnnotation::class.java)) continue
            val annotation = method.getAnnotation(DebotAnnotation::class.java)
            if (methodName == (annotation as DebotAnnotation).value) {
                try {
                    method.invoke(activity)
                } catch (e: InvocationTargetException) {
                    e.printStackTrace()
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                }

            }

        }
    }
}
